src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [Object.assign({}, state)];
+
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+    let newState = Object.assign({}, states[i]);
+
+    if (action.type === 'addProperties') {
+      newState = Object.assign(newState, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      for (let j = 0; j < action.keysToRemove.length; j++) {
+        delete newState[action.keysToRemove[j]];
+      }
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
+
+    states.push(newState);
+  }
```

Comment:
let's rewrite it with `for of` and `switch\case` (don't forget about the default case) to make your code more readable